sandbox/apparmor/apparmor.go

```diff
@@ -56,6 +57,221 @@ func ValidateNoAppArmorRegexp(s string) error {
+				break findCommonPrefix
+			}
+		}
+		commonPrefix = excludePatterns[0][:charInd+1]
+	}
+
+	// This loop will iterate over the length of the longest excludePattern
+	// (charInd = 1..size), generating an apparmor rule on each iteration
+	// for the corresponding subpatterns, understanding as such, the first
+	// (charInd+1) characters of the excludePatterns.
+	builder := &strings.Builder{}
+	for charInd := 1; charInd < size; charInd++ {
+
+		// This loop will group the subpatterns properly, generating the subpatterns map, where:
+		//     - the key would be the subpatternPrefix, considering as such the subpattern except
+		//       its last character (pattern[0:charInd]).
+		//     - the value would be the charset, which would be the subpattern last character
+		//       (pattern[charInd]). If several subpatterns share the same subpatternPrefix, the
+		//       charset would be a string including the last character of all those subpatterns.
+		subpatternPrefix := ""
+		subpatterns := map[string]string{}
+		for _, pattern := range excludePatterns {
+			// Handle unsupported cases
+			if (charInd < len(pattern)) && pattern[charInd] == '*' {
+				// Check if the excludePattern has a character different from '/' after a wildcard
+				if ((charInd + 1) < len(pattern)) && (pattern[charInd+1] != '/') {
+					return "", errors.New("exclude patterns does not support suffixes for now")
+				}
+				// Check if '*' is the first character after a common subpattern
+				for _, patt := range excludePatterns {
+					if (patt != pattern) && ((charInd) < len(patt)) && (pattern[:charInd] == patt[:charInd]) && (patt[charInd] != '*') {
+						return "", errors.New("first character after a common subpattern cannot be a wildcard")
+					}
+				}
+			}
+
+			// Skip patterns that are already finished, wildcards and slashes preceded by wildcards.
+			if (charInd >= len(pattern)) ||
+				(pattern[charInd] == '*') ||
+				((pattern[charInd] == '/') && (pattern[charInd-1] == '*')) {
+				continue
+			}
+			// Group subpatterns
+			subpatternPrefix = pattern[:charInd]
+			if charset, exists := subpatterns[subpatternPrefix]; !exists {
+				// Add the pattern if it didn't exist yet
+				subpatterns[subpatternPrefix] = string(pattern[charInd])
+			} else {
+				if !strings.Contains(charset, string(pattern[charInd])) {
+					// Two patterns only differ on the last character
+					subpatterns[subpatternPrefix] = charset + string(pattern[charInd])
+				}
+			}
+		}
+
+		// Write patterns
+		if len(subpatterns) == 0 {
+			continue
+		} else if len(subpatterns) == 1 {
+			res := fmt.Sprintf("%s%s[^%s]**%s\n", opts.Prefix, subpatternPrefix, subpatterns[subpatternPrefix], opts.Suffix)
+			builder.WriteString(res)
+		} else {
+			var aux []string
+			for prefix, charset := range subpatterns {
+				aux = append(aux, prefix[len(commonPrefix):]+"[^"+charset+"]")
+			}
+			res := fmt.Sprintf("%s%s{%s}**%s\n", opts.Prefix, commonPrefix, strings.Join(aux[:], ","), opts.Suffix)
```

Comment:
Do we need to format a string to add it to the builder? We can just
```go
builder.WriteString(opts.Prefix)
builder.WriteString(commonPrefix)
builder.WriteRune('{')
for i := range aux {
  if i > 0 {
    builder.WriteRune(',')
  }
  builder.WriteString(aux[i])
}
builder.WriteRune('}')
builder.WriteString(suffix)
```

This will be less wasteful on the Gc.

Reply:
It looks less readable and harder to see what is actually doing tbh. Do you think it would be worth the performance gain?

Reply:
I think that is the whole point of using the builder.

Reply:
Umm, I'm now wondering whether we need a builder at all. I retained this pattern from the original PR but maybe simple string concatenation is enough for this one. What do you think?

Reply:
TBH builder is preferred to any form of string concatenation.

Reply:
I think

Reply:
Fine for me, will do :)

Reply:
Done in https://github.com/snapcore/snapd/pull/13488/commits/5d3279b19f6c81d917f386a1d3985abccb317ff7